### Basics of TensorFlow

In [2]:
import tensorflow as tf

In [3]:
#Constants
con1 = tf.constant(6.5, tf.float32)

In [4]:
#Starting a Session to run the Computational Graph
sess = tf.Session()

In [5]:
#Printing out the values in session
print(sess.run(con1))

6.5


In [8]:
#PlaceHolders
ph1 = tf.placeholder(tf.float32)
ph2 = tf.placeholder(tf.float32)

In [9]:
addition = ph1 + ph2

In [11]:
print(sess.run(addition, {ph1: [1,2,3], ph2: [4,5,6]}))

[5. 7. 9.]


In [13]:
#Variables
var1 = tf.Variable([2.0], tf.float32)
var2 = tf.Variable([5.], tf.float32)

### Linear Model

In [15]:
W = tf.Variable([.3], tf.float32) #Weights
b = tf.Variable([-.3], tf.float32) #Bias
x = tf.placeholder(tf.float32) #Inputs

In [16]:
linear_model = W*x + b
init = tf.global_variables_initializer()

#New Session
sess_lm = tf.Session()
sess_lm.run(init)
print(sess_lm.run(linear_model, {x:[1,2,3,4]}))

[0.         0.3        0.6        0.90000004]


### Calculating Loss

In [17]:
y = tf.placeholder(tf.float32)
squared_deltas = tf.square(linear_model - y)

loss = tf.reduce_sum(squared_deltas)
print(sess_lm.run(loss, {x:[1,2,3,4], y:[0, -1, -2, -3]}))

23.66


#### We got 23.66 loss. So, we try to minimize that by using Optimizers provided by TensorFlow

In [18]:
optimizer = tf.train.GradientDescentOptimizer(0.01) #Learning Rate = 0.01
train = optimizer.minimize(loss)

sess_lm.run(init)

for i in range(1000):
    sess_lm.run(train, {x:[1,2,3,4], y:[0, -1, -2, -3]})
    
print(sess_lm.run([W,b]))

[array([-0.9999969], dtype=float32), array([0.9999908], dtype=float32)]
